File: LAB/Lab_1/Z4.ml
exception IllegalArgument of string

let militaryMinutes (hours: int) (minutes: int) (timeOfDay: string) =
  let buildString = fun (h: int) (m: int) -> Printf.sprintf "%d:%d" h m in

  let amFunc (h: int) (m: int) =
    if h = 12 then
      buildString 0 m
    else
      buildString h m
  in

  let pmFunc (h: int) (m: int) =
    if h = 12 then
      buildString h m
    else
      buildString (h + 12) m
  in

  if minutes > 59 || minutes < 0 then
    failwith "Minutes must belong to interval [0;59]"
  else if hours > 12 || hours < 0 then
    failwith "Hours must belong to interval [0;12]"
  else
    if timeOfDay = "PM" then
      pmFunc hours minutes
    else if timeOfDay = "AM" then
      amFunc hours minutes
    else
      failwith "Time of day must be AM or PM"

let timeA = militaryMinutes 5 12 "PM";;
let timeB = militaryMinutes 12 59 "AM";;
let timeC = militaryMinutes 12 42 "PM";;

Printf.printf "%s, %s, %s" timeA timeB timeC;;

let timeD = militaryMinutes 12 60 "AM";;

File: LAB/Lab_1/Z4.scala
def militaryMinutes(hours: Int, minutes: Int, timeOfDay: String): String  = {
    val buildString = (h: Int, m: Int) => h + ":" + m;

    def amFunc(h: Int, m: Int) = {
        if (h == 12) buildString(0, m);
        else buildString(h, m);
    }

    def pmFunc(h: Int, m: Int) = {
        if (h == 12) buildString(h, m);
        else buildString(h + 12, m);
    }

    if (minutes > 59 || minutes < 0) {
        throw new IllegalArgumentException("Minutes should be between 0 and 59.");
    } else if (hours > 12 || hours < 0) {
        throw new IllegalArgumentException("Hours should be between 0 and 12");
    }

    if (timeOfDay == "PM") {
        pmFunc(hours, minutes);
    } else if (timeOfDay == "AM") {
        amFunc(hours, minutes)
    } else {
        throw new IllegalArgumentException("Time of day should be AM or PM.");
    }
}

@main def main() = {
    println(militaryMinutes(5, 12, "PM"));
    println(militaryMinutes(12, 59, "AM"));
    println(militaryMinutes(12, 43, "PM"));
    println(militaryMinutes(12, 60, "AM"));
}